keps/sig-node/1664-node-ips/README.md

```diff
@@ -0,0 +1,465 @@
+
+For enhancements that make changes to code or processes/procedures in core
+Kubernetes i.e., [kubernetes/kubernetes], we require the following Release
+Signoff checklist to be completed.
+
+Check these off as they are completed for the Release Team to track. These
+checklist items _must_ be updated for the enhancement to be released.
+-->
+
+- [ ] Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] KEP approvers have approved the KEP status as `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+`Node.Status.Addresses` has unclear and inconsistent semantics, and has
+a note attached to it in the documentation warning you to not use it
+with strategic merge patch because of an incorrect but unfixable
+annotation.
+
+To resolve these problems, this KEP proposes deprecating the existing
+`Node.Status.Addresses` field and replacing it with two new fields,
+`Node.Status.Hostname` and `Node.Status.IPs`, with clearer and more
+consistent semantics.
+
+## Motivation
+
+### Problems
+
+#### API cleanliness
+
+`Node.Status.Addresses` is annotated with an incorrect strategic-merge
+patch annotation, and so it is not possible to use strategic merge patch
+to change that field without losing information. Although kubelet and
+cloud-controller-manager have been fixed to work around the problem, it
+is still annoying to have the incorrect annotation there, and it is
+possible that other components might attempt to update the field
+incorrectly in the future.
+
+#### IPv6 and Dual-stack problems
+
+Nodes must have node IPs corresponding to each IP family in use in the
+cluster. Eg, kube-proxy can only create rules for NodePorts for IPv4
+Services if there is an IPv4 node IP, and it can only create rules for
+NodePorts for IPv6 Services if there is an IPv6 node IP.
+
+When using an external cloud provider, cloud-controller-manager is
+responsible for setting `Node.Status.Addresses`, but it does not know
+whether the cluster it is running in is supposed to be single-stack
+IPv4, single-stack IPv6, or dual stack. If a node has both IPv4 and IPv6
```

Comment:
I wonder if it would be more bang for our buck to just add new node address types for v6 and call it a day. If we add a `InternalIPv6` and `ExternalIPv6`, users can configure the priority based on `--kubelet-preferred-address-types`. The strategic merge patch issue would still exist but I don't think it's a major issue since we have workarounds in both the kubelet and the cloud-controller-manager.

Reply:
Cloud-controller-manager sets `Node.Status.Addresses` directly, but it doesn't get any input from kubelet on how to do it. So to implement a `--preferred-address-types` arg in kubelet we need to improve kubelet/CCM communication as well. (Presumably via a Node annotation, which is how `--node-ip` gets passed.)

I should call this out more explicitly as a problem in the Motivation section.

Reply:
`--kubelet-preferred-address-types` is an apiserver flag btw.

Reply:
> I wonder if it would be more bang for our buck to just add new node address types for v6 and call it a day.

I was mostly only fiddling with `NodeAddressType` because if we're going to replace `Node.Status.Addresses` then we might as well try to fix *all* the problems with it at once. (Did I miss any?) If we aren't going replace it, I'm not sure it's super important to distinguish more address types. Calling globally-routable IPv6 addresses `InternalIP` is confusing but it *works* fine.

(And changing existing IPv6 addresses to `InternalIPv6` will immediately break existing code that only looks at `InternalIP` and `ExternalIP`. This is another reason why adding the new fields is nice; because we can still leave everything as it was in the old fields.)

Reply:
> (And changing existing IPv6 addresses to InternalIPv6 will immediately break existing code that only looks at InternalIP and ExternalIP. This is another reason why adding the new fields is nice; because we can still leave everything as it was in the old fields.)

Good point, my assumption is that this only breaks when going from IPv6 -> Dual Stack both of which are alpha features. 

> If we aren't going replace it, I'm not sure it's super important to distinguish more address types. Calling globally-routable IPv6 addresses InternalIP is confusing but it works fine.

The real benefit is allowing users to configure the preferred family. Setting the type for an IPv6 address to `InternalIP` works for IPv6 only. In dual-stack it falls apart since order matters and the order is up to the cloud provider. It's not really feasible for the cloud provider to know the preferred order of address I think. This should be up to the user.

Reply:
I wasn't aware of `apiserver --kubelet-preferred-address-type`. Need to think about how that fits into this.

This is basically for the case where the masters are on a separate network from the nodes and can't reach the "internal" IPs, right?

At first glance, it seems like we ought to be able to decide whether apiserver→node communication should be IPv4 or IPv6 based on whether the apiserver is advertising an IPv4 or an IPv6 IP to `kubernetes.default`. If node→apiserver communication is going to be over IPv6 then it seems to make sense to assume that apiserver→node communication should also be IPv6?